R/params.R

```diff
@@ -0,0 +1,355 @@
+
+knit_params_get <- function(input_lines, params) {
+  # check for recent enough knitr
+  if (packageVersion("knitr") < "1.10") {
+    stop("knitr >= 1.10 required to use rmarkdown params")
+  }
+  
+  # read the default parameters and extract them into a named list
+  knit_params <- mark_utf8(knitr::knit_params(input_lines))
+  default_params <- list()
+  for (param in knit_params) {
+    default_params[[param$name]] <- param$value
+  }
+  
+  # validate params passed to render
+  if (!is.null(params)) {
+
+    if (identical(params, "ask")) {
+      if (!interactive()) {
+        stop("render parameter configuration only allowed in an interactive environment")
+      }
+      
+      params <- knit_params_ask(input_lines = input_lines)
+      if (is.null(params)) {
+        stop("render parameter configuration canceled")
+      }
+    }
+    
+    # verify they are a list
+    if (!is.list(params) || (length(names(params)) != length(params))) {
+      stop("render params argument must be a named list")
+    }
+    
+    # verify that all parameters passed are also in the yaml
+    invalid_params <- setdiff(names(params), names(default_params))
+    if (length(invalid_params) > 0) {
+      stop("render params not declared in YAML: ",
+           paste(invalid_params, sep = ", "))
+    }
+  }
+  
+  # merge explicitly provided params with defaults
+  merge_lists(default_params, params, recursive = FALSE)
+}
+
+params_convert_to_ui <- function(inputControlFn, value) {
+  ## TODO: if long input, maybe truncate textInput values for display
+  if (identical(inputControlFn, shiny::fileInput)) {
+    NULL
+  } else if (identical(inputControlFn, shiny::textInput) &&
+             "POSIXct" %in% class(value)) {
+    ## FIXME: find the right conversion fn.
+    format(value)
+  } else {
+    ## A type/control that doesn't need special handling; just emit the value.
+    value
+  }
+}
+
+params_convert_from_ui <- function(inputControlFn, value, uivalue) {
+  if (identical(inputControlFn, shiny::fileInput)) {
+    uivalue$datapath
+  } else if (identical(inputControlFn, shiny::textInput) &&
+             "POSIXct" %in% class(value)) {
+    ## FIXME: This is how knitr does datetime conversion, but maybe we should use the local tz.
+    as.POSIXct(uivalue, tz="GMT")
+  } else {
+    ## A type/control that doesn't need special handling; just emit the value.
+    uivalue
+  }
+}
+
+params_get_input <- function(param) {
+  # Maps between value types and input: XXX
+  default_inputs <- list(
+      logical = "checkbox",
+      ## BUG: dateInput does not allow the user to not specify a value.
+      ##     https://github.com/rstudio/shiny/issues/896
+      Date = "date",
+      ## BUG: shiny does not support datetime selectors
+      ##     https://github.com/rstudio/shiny/issues/897
+      ##     we ask for string input for now.
+      POSIXct = "datetime",
```

Comment:
I agree that Shiny ought to have datetime input, but in general, don't feel like you need to limit yourself to the inputs that Shiny supports out of the box--if you want to bundle your own custom controls with rmarkdown that would be fine IMHO.